File: app/src/main/java/com/govnokoder/velotracker/MainActivity.java
package com.govnokoder.velotracker;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import com.google.android.material.navigation.NavigationView;
import com.google.android.material.tabs.TabLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatCheckBox;
import androidx.appcompat.widget.Toolbar;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.appcompat.app.AppCompatActivity;
import androidx.viewpager2.adapter.FragmentStateAdapter;
import androidx.viewpager2.widget.ViewPager2;

import com.google.android.material.tabs.TabLayoutMediator;
import com.govnokoder.velotracker.BL.Controller.TrainingController;
import com.govnokoder.velotracker.BL.ParcelableTraining;
import com.govnokoder.velotracker.ui.main.PageStart;
import com.govnokoder.velotracker.ui.main.ViewPagerAdapter;

public class MainActivity extends AppCompatActivity implements PageStart.onSomeEventListener {

    TabLayout tabLayout;

    private NavigationView navigationView;
    private DrawerLayout drawerLayout;

    ViewPager2 pager;
    private final int REQUEST_PERMISSION_ACCESS_FINE_LOCATION = 111;
    private final int REQUEST_PERMISSION_READ_PHONE_STATE = 333;

    boolean gpsEnabled;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        pager = (ViewPager2)findViewById(R.id.view_pager);
        FragmentStateAdapter pageAdapter = new ViewPagerAdapter(this);
        pager.setAdapter(pageAdapter);

        tabLayout = findViewById(R.id.tabs);

        TabLayoutMediator tabLayoutMediator= new TabLayoutMediator(tabLayout, pager, new TabLayoutMediator.TabConfigurationStrategy(){
            @Override
            public void onConfigureTab(@NonNull TabLayout.Tab tab, int position) {
                switch (position) {
                    case 0:
                        tab.setText(getText(R.string.start));
                        break;
                    case 1:
                        tab.setText(getText(R.string.history));
                        break;
                    case 2:
                        tab.setText(getText(R.string.statistic));
                        break;
                }
            }
        });
        tabLayoutMediator.attach();
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        drawerLayout = findViewById(R.id.drawer_layout);
        navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                //Установите слушателя, который будет получать уведомления при выборе пункта меню
                switch (item.getItemId()){
                    case R.id.nav_start:
                        tabLayout.selectTab(tabLayout.getTabAt(0));
                        break;
                    case R.id.nav_history:
                        tabLayout.selectTab(tabLayout.getTabAt(1));
                        break;
                    case R.id.nav_statistics:
                        tabLayout.selectTab(tabLayout.getTabAt(2));
                        break;
                    case R.id.nav_settings:
                        startActivity(new Intent(getApplicationContext(), SettingsActivity.class));
                        //TODO доделать настройки
                        break;
                }
                drawerLayout.closeDrawers();
                return false;
            }
        });
        toolbar = findViewById(R.id.toolbar);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                drawerLayout.open();
            }
        });
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        boolean allowed = false;
        String currentPer = permissions[0];
        allowed = grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED;
        if(allowed){
            startTraining();
            return;
        }
        if (currentPer.equals(Manifest.permission.ACCESS_FINE_LOCATION) && !shouldShowRequestPermissionRationale(Manifest.permission.ACCESS_FINE_LOCATION)){
            dialogOpenPermissionsSettings(getString(R.string.dialog_open_sett_access_fine_location_text));
        }
        if (currentPer.equals(Manifest.permission.READ_PHONE_STATE) && !shouldShowRequestPermissionRationale(Manifest.permission.READ_PHONE_STATE)){
            dialogOpenPermissionsSettings(getString(R.string.dialog_open_sett_read_phone_state_text));
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == AppConstants.GPS_REQUEST) {
                //isGPS = true; // flag maintain before get location
            }
        }
    }

    private void startActivityTraining(){
        Intent intent = new Intent(getApplicationContext(), TrainingActivity.class);
        startActivity(intent);
        finish();
    }

    @Override
    public void openLastTraining() {
        int position = new TrainingController(this).LoadTrainingsData().size() - 1;
        if(position >= 0) {
            tabLayout.selectTab(tabLayout.getTabAt(1));
            Intent intent = new Intent(this, LookTraining.class);
            intent.putExtra("INDEX", position);
            startActivity(intent);
        }
    }

    @Override
    public void openStatistic() {
        tabLayout.selectTab(tabLayout.getTabAt(2));
    }

    private void dialogOpenPermissionsSettings(String text){
        AlertDialog dialog = new AlertDialog.Builder(this).create();
        ConstraintLayout cl  = (ConstraintLayout)getLayoutInflater().inflate(R.layout.dialog_open_settings, null);
        TextView textView = (TextView) cl.getViewById(R.id.textView);
        textView.setText(text);
        cl.getViewById(R.id.openSettB).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS, Uri.parse("package:" + getPackageName())));
                dialog.dismiss();
            }
        });
        dialog.setView(cl);
        dialog.show();
    }

    private void dialogBatteryOptimizationSettings(){
        //если таких настроек не существует, они создаются
        SharedPreferences sharedPreferences = getSharedPreferences(getString(R.string.settings_checkboxes), MODE_PRIVATE);
        if(sharedPreferences.contains(getString(R.string.dialog_battery_optimization_settings_checkbox_key))){
            if(sharedPreferences.getBoolean(getString(R.string.dialog_battery_optimization_settings_checkbox_key), false)) {
                //если стоит больше не показывать
                startActivityTraining();
                return;
            }
        }

        final boolean[] isInSettings = {false};
        //добавляю настройку и потом вызываю диалог
        sharedPreferences.edit().putBoolean(getString(R.string.dialog_battery_optimization_settings_checkbox_key), false).apply();

        AlertDialog dialog = new AlertDialog.Builder(this).create();
        ConstraintLayout cl  = (ConstraintLayout)getLayoutInflater().inflate(R.layout.dialog_open_settings_with_checkbox, null);
        TextView textView = (TextView) cl.getViewById(R.id.textView);
        textView.setText(getString(R.string.dialog_battery_optimization_text));
        cl.getViewById(R.id.openSettB).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                isInSettings[0] = true;
                startActivity(new Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS));
                dialog.dismiss();
            }
        });
        AppCompatCheckBox compatCheckBox = (AppCompatCheckBox) cl.getViewById(R.id.checkbox);
        dialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
            @Override
            public void onDismiss(DialogInterface dialog) {
                if(compatCheckBox.isChecked()){
                    SharedPreferences sharedPreferences = getSharedPreferences(getString(R.string.settings_checkboxes), MODE_PRIVATE);
                    sharedPreferences.edit().putBoolean(getString(R.string.dialog_battery_optimization_settings_checkbox_key), true).apply();
                }
                if(!isInSettings[0]) {
                    startActivityTraining();
                    return;
                }
            }
        });
        dialog.setView(cl);
        dialog.show();
    }

    @Override
    public void startTraining() {
        //тут переопределяю метод из интерфейса который в фрагменте
        boolean accessFineLocation = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        boolean readPhoneState = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED;
        LocationManager locationManager = (LocationManager) getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
        gpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);

        if(!accessFineLocation){
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, REQUEST_PERMISSION_ACCESS_FINE_LOCATION);
            return;
        }

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q){
            if (!readPhoneState) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_PHONE_STATE}, REQUEST_PERMISSION_READ_PHONE_STATE);
                return;
            }
        }

        if(!gpsEnabled) {
            GpsUtils gpsUtils = new GpsUtils(this);
            gpsUtils.turnGPSOn(new GpsUtils.onGpsListener() {
                @Override
                public void gpsStatus(boolean isGPSEnable) {
                    if(isGPSEnable){
                        startTraining();
                    }
                }
            });
        }

        if(!gpsEnabled){
            return;
        }
        dialogBatteryOptimizationSettings();
    }
}

File: app/src/main/java/com/coursework/velotracker/MainActivity.kt
package com.coursework.velotracker

import android.Manifest
import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.location.LocationManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.MenuItem
import android.view.View
import android.view.View.VISIBLE
import android.view.ViewAnimationUtils
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.appcompat.widget.*
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.coursework.velotracker.BL.Controller.TrainingController
import com.coursework.velotracker.databinding.ActivityMainBinding
import com.coursework.velotracker.ui.fragments.main.PageStart
import com.coursework.velotracker.ui.fragments.main.ViewPagerAdapter
import com.google.android.material.navigation.NavigationView
import com.google.android.material.tabs.TabLayoutMediator

class MainActivity : AppCompatActivity(), PageStart.OnSomeEventListener, NavigationView.OnNavigationItemSelectedListener {

    private lateinit var binding: ActivityMainBinding
    private lateinit var header: View
    private lateinit var themeBtn: AppCompatImageButton
    private val REQUEST_PERMISSION_ACCESS_FINE_LOCATION = 111

    private var gpsEnabled = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private fun changeTheme(){
        val sharedPreferences = getSharedPreferences("theme", MODE_PRIVATE)
        if(sharedPreferences.getBoolean("dark", false)){
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            sharedPreferences.edit().putBoolean("light", true).apply()
            sharedPreferences.edit().putBoolean("dark", false).apply()
            themeBtn.foreground = getDrawable(R.drawable.ic_light_mode_foreground)
        }
        else if(sharedPreferences.getBoolean("light", false)){
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            sharedPreferences.edit().putBoolean("dark", true).apply()
            sharedPreferences.edit().putBoolean("light", false).apply()
            themeBtn.foreground = getDrawable(R.drawable.ic_dark_mode_foreground)
        }
        else{
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            sharedPreferences.edit().putBoolean("light", true).apply()
            sharedPreferences.edit().putBoolean("dark", false).apply()
            themeBtn.foreground = getDrawable(R.drawable.ic_light_mode_foreground)
        }
    }


    override fun onStart() {
        super.onStart()
        initFunc()
    }

    private fun initFunc(){
        createViewPager()
        setSupportActionBar(binding.toolbar)
        binding.toolbar.setNavigationOnClickListener { binding.drawerLayout.open() }
        binding.navView.setNavigationItemSelectedListener(this)
        header = binding.navView.getHeaderView(0)
        themeBtn = header.findViewById<AppCompatImageButton>(R.id.themeBtn)
        themeBtn.setOnClickListener {changeTheme()}
        setCurrentTheme()
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    private fun setCurrentTheme(){
        val sharedPreferences = getSharedPreferences("theme", MODE_PRIVATE)
        if(sharedPreferences.contains("dark") && sharedPreferences.getBoolean("dark", false)){
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            sharedPreferences.edit().putBoolean("dark", true).apply()
            sharedPreferences.edit().putBoolean("light", false).apply()
            themeBtn.foreground = getDrawable(R.drawable.ic_light_mode_foreground)
        }
        else if(sharedPreferences.contains("light") && sharedPreferences.getBoolean("light", false)){
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            sharedPreferences.edit().putBoolean("dark", false).apply()
            sharedPreferences.edit().putBoolean("light", true).apply()
            themeBtn.foreground = getDrawable(R.drawable.ic_dark_mode_foreground)
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.nav_start -> binding.tabs.selectTab(binding.tabs.getTabAt(0))
            R.id.nav_history -> binding.tabs.selectTab(binding.tabs.getTabAt(1))
            R.id.nav_statistics -> binding.tabs.selectTab(binding.tabs.getTabAt(2))
            R.id.nav_about_program -> startActivity(
                Intent(
                    applicationContext,
                    AboutProgramActivity::class.java
                )
            )
        }
        binding.drawerLayout.closeDrawers()
        return false
    }

    private fun createViewPager(){
        val pagerAdapter = ViewPagerAdapter(this)
        binding.viewPager.adapter = pagerAdapter
        val tabLayoutMediator = TabLayoutMediator(binding.tabs, binding.viewPager) { tab, position ->
            when (position) {
                0 -> tab.text = getText(R.string.start)
                1 -> tab.text = getText(R.string.history)
                2 -> tab.text = getText(R.string.statistic)
            }
        }
        tabLayoutMediator.attach()
    }

    @SuppressLint("ResourceType")
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        var allowed = false
        val currentPer = permissions[0]
        allowed = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
        if(allowed){
            startTraining()
            return
        }
        if(currentPer == Manifest.permission.ACCESS_FINE_LOCATION && !shouldShowRequestPermissionRationale(
                Manifest.permission.ACCESS_FINE_LOCATION
            )) {
            dialogOpenPermissionsSettings(getString(R.string.dialog_open_sett_access_fine_location_text))
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if(resultCode == Activity.RESULT_OK && requestCode == AppConstants.GPS_REQUEST){
            //isGPS = true
        }
    }

    private fun startActivityTraining() {
        val intent = Intent(applicationContext, TrainingActivity::class.java)
        startActivity(intent)
        finish()
    }

    override fun openLastTraining() {
        val position: Int = TrainingController(this).loadTrainings().size - 1
        if (position >= 0) {
            binding.tabs.selectTab(binding.tabs.getTabAt(1))
            val intent = Intent(this, LookTraining::class.java)
            intent.putExtra("INDEX", position)
            startActivity(intent)
        }
    }

    override fun openStatistic() {
        binding.tabs.selectTab(binding.tabs.getTabAt(2))
    }

    private fun dialogOpenPermissionsSettings(text: String) {
        val dialog: AlertDialog = AlertDialog.Builder(this).create()
        val cl = layoutInflater.inflate(R.layout.dialog_open_settings, null) as ConstraintLayout
        val textView = cl.getViewById(R.id.textView) as TextView
        textView.text = text
        cl.getViewById(R.id.openSettB).setOnClickListener {
            startActivity(
                Intent(
                    Settings.ACTION_APPLICATION_DETAILS_SETTINGS,
                    Uri.parse("package:$packageName")
                )
            )
            dialog.dismiss()
        }
        dialog.setView(cl)
        dialog.show()
    }

    private fun dialogBatteryOptimizationSettings() {
        //если таких настроек не существует, они создаются
        val sharedPreferences = getSharedPreferences(
            getString(R.string.settings_checkboxes),
            MODE_PRIVATE
        )
        if (sharedPreferences.contains(getString(R.string.dialog_battery_optimization_settings_checkbox_key))) {
            if (sharedPreferences.getBoolean(
                    getString(R.string.dialog_battery_optimization_settings_checkbox_key),
                    false
                )) {
                //если стоит больше не показывать
                startActivityTraining()
                return
            }
        }
        val isInSettings = booleanArrayOf(false)
        //добавляю настройку и потом вызываю диалог
        sharedPreferences.edit().putBoolean(
            getString(R.string.dialog_battery_optimization_settings_checkbox_key),
            false
        ).apply()
        val dialog = AlertDialog.Builder(this).create()
        val cl = layoutInflater.inflate(R.layout.dialog_open_settings_with_checkbox, null) as ConstraintLayout
        val textView = cl.getViewById(R.id.textView) as TextView
        textView.text = getString(R.string.dialog_battery_optimization_text)
        cl.getViewById(R.id.openSettB).setOnClickListener {
            isInSettings[0] = true
            startActivity(Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS))
            dialog.dismiss()
        }
        val compatCheckBox = cl.getViewById(R.id.checkbox) as AppCompatCheckBox
        dialog.setOnDismissListener(DialogInterface.OnDismissListener {
            if (compatCheckBox.isChecked) {
                val sharedPreferences = getSharedPreferences(
                    getString(R.string.settings_checkboxes),
                    MODE_PRIVATE
                )
                sharedPreferences.edit().putBoolean(
                    getString(R.string.dialog_battery_optimization_settings_checkbox_key),
                    true
                ).apply()
            }
            if (!isInSettings[0]) {
                startActivityTraining()
                return@OnDismissListener
            }
        })
        dialog.setView(cl)
        dialog.show()
    }

    override fun startTraining() {
        val accessFineLocation = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED
        val locationManager = applicationContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
        gpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
        if (!accessFineLocation) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), REQUEST_PERMISSION_ACCESS_FINE_LOCATION)
            return
        }
        if (!gpsEnabled) {
            val gpsUtils = GpsUtils(this)
            gpsUtils.turnGPSOn(object : GpsUtils.OnGpsListener {
                override fun gpsStatus(isGPSEnable: Boolean) {
                    if (isGPSEnable) {
                        startTraining()
                    }
                }
            })
        }
        if (!gpsEnabled) {
            return
        }
        dialogBatteryOptimizationSettings()
    }
}
